README.md

```diff
@@ -2,4 +2,61 @@
 
 A multi-threaded implementation of a PromQL Query Engine based on the [Volcano/Iterator model](https://paperhub.s3.amazonaws.com/dace52a42c07f7f8348b08dc2b186061.pdf).
 
-The project is currently under development.
+The project is currently under active development.
+
+## Design
+
+At the beggining of a PromQL query execution, the query engine computes a physical plan consisting of multiple independent operators, each responsible for calculating one part of the query expression.
+
+Operators are assembled in a tree-like structure with every operator calling `Next()` on its dependants until there is no more data to be returned. The result of the `Next()` function is a _column vector_  (also called a _step vector_) with elements in the vector representing samples with the same timestamp from different time series.
+
+<p align="center">
+  <img src="./assets/design.png" />
```

Comment:
Hm, I am still unsure how rate/over_time fits into this operator model as they can be fully parallelized across series (series independent). I can see it's not yet implemented/used in the current implementation.

We can leave it for now, but let's add some todo/mention future work (:  Essentially e.g rate or sum_over_time requires multiple "steps" which might be different than root steps!

Reply:
Right now we can push rates into selectors so that the output is still a step vector. So the matrix selector would have a `FuncCall` that it executes over selected samples and produces a single point that is then sent as part of a vector to the next operator. 

Right now it's not implemented because there was too much refactoring and changing of interfaces, so it became cumbersome to keep everything up to date :)

Reply:
This assumes function call is on lowest level?

Reply:
AFAIK this is the only place where a function that returns multiple samples (like rate, sum_over_time, etc.) can be used.